src/status_im/common/biometric/events.cljs

```diff
@@ -1,120 +1,62 @@
-
-(rf/defn show-message
-  {:events [:biometric/show-message]}
-  [_ error]
-  (let [code    (ex-cause error)
-        content (if (#{::biometrics/not-enrolled
-                       ::biometrics/not-available}
+(defn show-message
+  [_ [code]]
+  (let [content (if (#{:biometrics/not-enrolled-error
+                       :biometrics/not-available-error}
                      code)
                   (i18n/label :t/grant-face-id-permissions)
                   (i18n/label :t/biometric-auth-error {:code code}))]
-    {:effects.utils/show-popup
-     {:title   (i18n/label :t/biometric-auth-login-error-title)
-      :content content}}))
-
-
-(re-frame/reg-fx
- :biometric/authenticate
- (fn [{:keys [on-success on-fail prompt-message]}]
-   (-> (biometrics/authenticate
-        {:prompt-message          (or prompt-message (i18n/label :t/biometric-auth-reason-login))
-         :fallback-prompt-message (i18n/label
-                                   :t/biometric-auth-login-ios-fallback-label)
-         :cancel-button-text      (i18n/label :t/cancel)})
-       (.then (fn [not-canceled?]
-                (when (and on-success not-canceled?)
-                  (on-success))))
-       (.catch (fn [err]
-                 (when on-fail
-                   (on-fail err)))))))
-
-(rf/defn authenticate
-  {:events [:biometric/authenticate]}
-  [_ opts]
-  {:biometric/authenticate opts})
-
-(rf/reg-event-fx
- :biometric/on-enable-success
- (fn [{:keys [db]} [password]]
-   (let [key-uid (get-in db [:profile/profile :key-uid])]
-     {:db       (assoc db :auth-method constants/auth-method-biometric)
-      :dispatch [:keychain/save-password-and-auth-method
-                 {:key-uid         key-uid
-                  :masked-password password}]})))
-
-(rf/reg-event-fx
- :biometric/enable
- (fn [_ [password]]
-   {:dispatch [:biometric/authenticate
-               {:on-success #(rf/dispatch [:biometric/on-enable-success password])
-                :on-fail    #(rf/dispatch [:biometric/show-message %])}]}))
-
-(rf/reg-event-fx
- :biometric/disable
- (fn [{:keys [db]}]
-   (let [key-uid (get-in db [:profile/profile :key-uid])]
-     {:db                           (assoc db :auth-method constants/auth-method-none)
-      :keychain/clear-user-password key-uid})))
+    {:fx [[:effects.utils/show-popup
+           {:title   (i18n/label :t/biometric-auth-login-error-title)
+            :content content}]]}))
+
+(schema/=> show-message events-schema/?show-message)
+(rf/reg-event-fx :biometric/show-message show-message)
+
+(defn authenticate
+  [{:keys [db]} [opts]]
+  (let [pending? (get-in db [:biometrics :auth-pending?])]
+    ;;NOTE: prompting biometric check while another one is pending triggers error
+    (when-not pending?
+      {:db (assoc-in db [:biometrics :auth-pending?] true)
+       :fx [[:effects.biometric/authenticate
+             (assoc opts :on-done #(rf/dispatch [:set-in [:biometrics :auth-pending?] false]))]]})))
```

Comment:
is it possible to avoid the use of set-in? can be a follow up too 👍

Reply:
will change it here. Wasn't aware at the time `set`/`set-in` were deprecated 👍

Reply:
nice one, yeah probably we should do a search within status-im folder for this pattern and clean it up so others stop adding this pattern too